src/dune_console/dune_console.ml

```diff
@@ -131,26 +143,74 @@ module Backend = struct
 
   let spawn_thread = Fdecl.create Dyn.opaque
 
-  let threaded (module Base : S) : (module S) =
+  (** Backends may have access to an internal state. This type is used by the
+      [Threaded] backends. *)
+  type state =
+    { messages : User_message.t Queue.t
+    ; mutable finish_requested : bool
+    ; mutable finished : bool
+    ; mutable status_line : User_message.Style.t Pp.t option
+    }
+
+  (** [Threaded] is the interface for user interfaces that are rendered in a
+      separate thread. It can be used to directly construct a threaded backend.
+
+      This module interface will be used by both the current console UI and the
+      upcoming terminal UI (TUI).
+
+      Notice that the TUI can also react to keyboard input, so it has the
+      additional [handle_user_events] call. *)
+  module type Threaded = sig
+    (** [start] is called by the main thread to start broadcasting the user
+        interface. Any initial setup should be performed here. *)
+    val start : unit -> unit
+
+    (** [render state] is called by the main thread to render the current state
+        of the user interface. *)
+    val render : state -> unit
+
+    (** [handle_user_events ~now ~time_budget mutex] is called by the main
+        thread to handle user events such as keypresses. The function should
+        return the time at which the next event should be handled. A [mutex] is
+        provided in order to lock the state of the UI. [time_budget] indicates
+        an amount of time that should be spent in this function. This is useful
+        for things like waiting for user input. *)
```

Comment:
The expectations we have from an implementation is a little unclear. Should the function spend `<= time_budget` time? Or maybe `= time_budget` time? Is it OK to exceed it slightly (e.g. when the `time_budget = 0.0`)?

Reply:
It's ok to be a little imprecise. The only consequence would be that we render slightly faster or slightly slower than the desired rendering rate (60 fps).

Reply:
All of this is now clarified in the code